### PR TITLE
[Build] Generate ObjC compatibility header for library Swift targets

### PR DESCRIPTION
**Pull request by aciidb0mb3r**
This will generate the ObjC header for library Swift targets so they can be
used in other ObjC targets. This is only done for macOS.

<rdar://problem/51780725>

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
@swift-ci smoke test

File: Sources/Build/BuildPlan.swift [hunk 1]
```diff
@@ -575,6 +592,37 @@ public final class SwiftTargetBuildDescription {
         return args
     }
 
+    /// Returns true if ObjC header should be emitted.
+    private var shouldEmitObjcHeader: Bool {
```
Comment (neonichu) [inline Sources/Build/BuildPlan.swift]: Maybe we should call this `shouldEmitObjCCompatibilityHeader` to make it more clear what this is for?

File: Tests/BuildTests/BuildPlanTests.swift [hunk 2]
```diff
@@ -1481,6 +1481,186 @@ final class BuildPlanTests: XCTestCase {
                 """), contents)
         }
     }
+
+    func testObjCHeader1() throws {
```
Comment (neonichu) [inline Tests/BuildTests/BuildPlanTests.swift]: Can we put a comment here what the test is for or make the name more specific?

**Comment by aciidb0mb3r** [comment]
@swift-ci smoke test

**Comment by aciidb0mb3r** [comment]
@swift-ci smoke test

**Comment by MarvinNazari** [comment]
@aciidb0mb3r: Is this released in Xcode 11.2? Because I'm still seeing the issue. I have a sample if it helps?

**Comment by aciidb0mb3r** [comment]
No, this change is not present in Xcode 11.2

**Comment by tomangistalis** [comment]
Is this going to be released with Xcode 11.3 or we have to wait for Swift 5.2?

**Comment by jlubeck** [comment]
How about Xcode 11.3?

**Comment by gurhub** [comment]
Maybe with 11.4?

**Comment by roman-dzieciol** [comment]
@aciidb0mb3r was there a particular reason for supporting only macOS? Would be great to have ObjC support for iOS too.

**Comment by oonoo** [comment]
Is there an update on when this is released / available in Xcode?

**Comment by wooster** [comment]
> This is only done for macOS.

@aciidb0mb3r Any reason not to do this for iOS, as well?

**Comment by SamuelZhaoY** [comment]
I wish this could be supported in iOS as well，currently I facing cases like delivering Swift packages to ObjC based App target.  Without able to directly import the autogenerated swift header in ObjC, it will be very painful to add a Swift adaptor layer in the app.

**Comment by devedup** [comment]
When can we have this for iOS? Thanks

**Comment by devedup** [comment]
This is working in iOS project on Xcode 11.4.1 for me.

I've tested it on an old Objective-C project and successfully can use Swift Package manager with it. Here is what I did: 

1. Swift Packages - add package dependency, as normal to include my package
2. In an Objective-C .m file, include the module like so:
`@import YourSwiftPackageName;`
3. Ensure in your swift package that you have used `@objc` to expose Swift to your Objective-C
4. Ensure you have used `public` to expose your Swift in your package to your main project, as it's in another module. 
5. Ensure your Swift Package is listed in your targets `'Frameworks, Libraries and Embedded Content'.` I note this because I have 2 targets - and it only added them to one of my targets, not both, so I kept getting Module not found errors.

**Comment by juliensechaud** [comment]
I'm able to add a SwiftPackage into MainTarget **Objective-C.m** file 
But no chance when trying to add it to my **Objective-C.h** file.
Any idea to solve this ?

**Comment by gistya** [comment]
Why is this only done for macOS?

**Pull request merged**